server/src/main/java/org/opensearch/common/logging/DeprecatedMessage.java

```diff
@@ -58,4 +74,12 @@ public DeprecatedMessage(String key, String xOpaqueId, String messagePattern, Ob
         }
         return builder.immutableMap();
     }
+
+    public boolean isAlreadyLogged() {
+        if (keys.contains(this.keyWithXOpaqueId)) {
+            return true;
+        }
+        keys.add(this.keyWithXOpaqueId);
```

Comment:
Is this safe on an unsynchronized HashSet? I'm assuming this method can be accessed concurrently from any number of threads. It's likely fine for the de-duplication logic to be best effort and a duplicate log entry emitted due to a race is not a big deal, but I'm more concerned about undefined behavior now or in future JDK versions if the underlying data structure is modified in the middle of a `contains()` invocation.

Reply:
Definitely agree! We can use the SynchronizedSet, the only downside might be performance since the collection will be locked by a single thread at a time. If that is not a concern, we can make this synchronized. WDYT?

Reply:
I think ConcurrentHashMap will give the concurrency guarantees while also allowing lock-free reads. Something like:

```
private final Set<String> keys = ConcurrentHashMap.newKeySet();

...

public boolean isAlreadyLogged() {
    return !keys.add(keyWithXOpaqueId);
}
```

Under the hood it looks like the KeySetView of ConcurrentHashMap is using the same method as `putIfAbsent` in its `add` method. That should be lock-free in the case that the key is present, and only incur any synchronization overhead in the case that it is newly adding a key.

Reply:
Sorry for the delay! That makes sense. I have updated the implementation to use the keySet() from ConcurrentHashMap.